File: desktop/src/models/task.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ItemStatus {
    Scheduled,
    Active,
    Completed,
    Cancelled,
}

impl ItemStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            ItemStatus::Scheduled => "Scheduled",
            ItemStatus::Active => "Active",
            ItemStatus::Completed => "Completed",
            ItemStatus::Cancelled => "Cancelled",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s {
            "Active" => ItemStatus::Active,
            "Completed" => ItemStatus::Completed,
            "Cancelled" => ItemStatus::Cancelled,
            _ => ItemStatus::Scheduled,
        }
    }

    pub fn is_terminal(&self) -> bool {
        matches!(self, ItemStatus::Completed | ItemStatus::Cancelled)
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum Priority {
    P1,
    P2,
    P3,
    P4,
}

impl Priority {
    pub fn as_str(&self) -> &'static str {
        match self {
            Priority::P1 => "P1",
            Priority::P2 => "P2",
            Priority::P3 => "P3",
            Priority::P4 => "P4",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s {
            "P1" => Priority::P1,
            "P2" => Priority::P2,
            "P3" => Priority::P3,
            _ => Priority::P4,
        }
    }

    pub fn label(&self) -> &'static str {
        match self {
            Priority::P1 => "P1 — Urgent",
            Priority::P2 => "P2 — High",
            Priority::P3 => "P3 — Medium",
            Priority::P4 => "P4 — Low",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum TaskType {
    Sequential,
    Parallel,
}

impl TaskType {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskType::Sequential => "Sequential",
            TaskType::Parallel => "Parallel",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s {
            "Parallel" => TaskType::Parallel,
            _ => TaskType::Sequential,
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Task {
    pub id: String,
    pub owner_id: String,
    pub project_id: Option<String>,
    pub parent_id: Option<String>,
    pub title: String,
    pub description: Option<String>,
    pub task_type: TaskType,
    pub weight: f64,
    pub priority: Priority,
    pub effective_priority: Priority,
    pub status: ItemStatus,
    pub start_at: String,
    pub due_at: String,
    pub completed_at: Option<String>,
    pub sort_order: i32,
    pub progress: f64,
    pub completed_child_count: i32,
    pub total_child_count: i32,
    pub next_child_title: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    #[serde(default)]
    pub children: Vec<Task>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateTaskPayload {
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    pub task_type: String,
    pub weight: f64,
    pub priority: String,
    pub start_at: String,
    pub due_at: String,
    pub sort_order: i32,
}

#[derive(Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateTaskPayload {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub weight: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub priority: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due_at: Option<String>,
}

File: openplan_mobile/shared/src/commonMain/kotlin/com/example/openplan_mobile/data/models/Task.kt
package com.example.openplan_mobile.data.models

import kotlinx.serialization.Serializable

@Serializable
data class Task(
    val id: String,
    val ownerId: String,
    val projectId: String? = null,
    val parentId: String? = null,
    val title: String,
    val description: String? = null,
    val taskType: String,
    val weight: Float = 1f,
    val priority: String,
    val effectivePriority: String,
    val status: String,
    val startAt: String,
    val dueAt: String,
    val completedAt: String? = null,
    val sortOrder: Int = 0,
    val progress: Double = 0.0,
    val completedChildCount: Int = 0,
    val totalChildCount: Int = 0,
    val nextChildTitle: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val children: List<Task> = emptyList()
) {
    val isCompleted get() = status == "Completed"
    val isCancelled get() = status == "Cancelled"
    val isTerminal get() = isCompleted || isCancelled
    val hasChildren get() = children.isNotEmpty() || totalChildCount > 0
    val isSequential get() = taskType == "Sequential"
}

@Serializable
data class CreateTaskRequest(
    val title: String,
    val description: String? = null,
    val projectId: String? = null,
    val parentId: String? = null,
    val taskType: String = "Parallel",
    val weight: Float = 1f,
    val priority: String = "P4",
    val startAt: String = "0001-01-01T00:00:00+00:00",
    val dueAt: String = "0001-01-01T00:00:00+00:00",
    val sortOrder: Int = 0
)

@Serializable
data class UpdateTaskRequest(
    val title: String? = null,
    val description: String? = null,
    val projectId: String? = null,
    val taskType: String? = null,
    val weight: Float? = null,
    val priority: String? = null,
    val status: String? = null,
    val startAt: String? = null,
    val dueAt: String? = null,
    val sortOrder: Int? = null
)
